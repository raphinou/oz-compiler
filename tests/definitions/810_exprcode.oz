% Pattern matchin, recursive calls, guards,...
local
   ExprCode
in

proc {ExprCode Expr Cin Cout Nin Nout}
   case Expr
   of [E1 plus E2] then C1 N1 C2 N2 in
     C1=plus|Cin
     N1=Nin+1
     {ExprCode E2 C1 C2    N1 N2}
     {ExprCode E1 C2 Cout  N2 Nout}
   [] I andthen {IsInt I} then
     Cout=push(I)|Cin
     Nout=Nin+1
   end
end

local C N in
   {ExprCode [[3 plus 3] plus 2] nil C 0 N}
   {Show C}
   {Show N}
end
end
