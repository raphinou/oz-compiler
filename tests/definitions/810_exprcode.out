[push(3) push(3) plus push(2) plus]
5
